src/metabase/sync_database/analyze.clj

```diff
@@ -157,86 +59,99 @@
-                    field-percent-urls-fn (partial driver/default-field-percent-urls driver)
-                    calculate-row-count?  true}}]
-  (fn [driver table new-field-ids]
-    (let [driver (assoc driver :field-avg-length field-avg-length-fn, :field-percent-urls field-percent-urls-fn)]
-      {:row_count (when calculate-row-count? (u/try-apply table-row-count table))
-       :fields    (for [{:keys [id] :as field} (table/fields table)]
-                    (let [new-field? (contains? new-field-ids id)]
-                      (cond->> {:id id}
-                               (test-for-cardinality? field new-field?) (test:cardinality-and-extract-field-values field)
-                               new-field?                               (test:new-field driver field))))})))
-
-(defn generic-analyze-table
-  "An implementation of `analyze-table` using the defaults (`default-field-avg-length` and `field-percent-urls`)."
-  [driver table new-field-ids]
-  ((make-analyze-table driver) driver table new-field-ids))
-
-
+(defn- percent-valid-urls
+  "Recursively count the values of non-nil values in VS that are valid URLs, and return it as a percentage."
+  [vs]
+  (loop [valid-count 0, non-nil-count 0, [v & more :as vs] vs]
+    (cond (not (seq vs)) (if (zero? non-nil-count) 0.0
+                             (float (/ valid-count non-nil-count)))
+          (nil? v)       (recur valid-count non-nil-count more)
+          :else          (let [valid? (and (string? v)
+                                           (u/is-url? v))]
+                           (recur (if valid? (inc valid-count) valid-count)
+                                  (inc non-nil-count)
+                                  more)))))
+
+(defn field-avg-length
+  "Default implementation of optional driver fn `field-avg-length` that calculates the average length in Clojure-land via `field-values-lazy-seq`."
+  [values]
+  (let [field-values       (filter identity values)
+        field-values-count (count field-values)]
+    (if (zero? field-values-count)
+      0
+      (int (math/round (/ (->> field-values
+                               (map str)
+                               (map count)
+                               (reduce +))
+                          field-values-count))))))
+
+(defn field-fingerprint [driver table field]
+  (let [values      (->> (driver/field-values-lazy-seq driver field)
+                         (take driver/max-sync-lazy-seq-results))
+        field-id    (:id field)
+        fingerprint {:base_type               (:base_type field)
+                     :is_pk                   (isa? (:special_type field) :type/PK)
+                     :is_fk                   (isa? (:special_type field) :type/FK)
+                     :cardinality             (count (distinct values))
+                     :field_percent_urls      (percent-valid-urls values)
+                     :field_percent_json      (if (values-are-valid-json? values) 100 0)
```

Comment:
this and `field_percent_email` should probably actually unwind these functions and compute the actual %, vs picking an extreme of 0 or 100